haystack/nodes/preprocessor/splitter.py

```diff
@@ -0,0 +1,738 @@
+        split_regex = split_regex if split_regex is not None else self.split_regex
+        split_length = split_length if split_length is not None else self.merger.window_size
+        split_overlap = split_overlap if split_overlap is not None else self.merger.window_overlap
+        max_chars = max_chars if max_chars is not None else self.max_chars
+        add_page_number = add_page_number if add_page_number is not None else self.add_page_number
+
+        self._validate_split_params(
+            split_by=split_by,
+            split_regex=split_regex,
+            split_length=split_length,
+            split_overlap=split_overlap,
+            max_chars=max_chars,
+            max_tokens=max_tokens,
+        )
+
+        if split_by == "token" and not self.tokenizer:
+            raise ValueError(
+                "If you set split_by='token', you must give a value to 'tokenizer_model'. "
+                "Use the same model you're using for your Reader."
+            )
+
+        if split_by == "character":
+            splitter_function = lambda text: (text, [0] * len(text)) if text != "" else ([""], [0])
+
+        elif split_by == "regex":
+            if not split_regex:
+                raise ValueError("If 'split_by' is set to 'regex', you must give a value to 'split_regex'.")
+            splitter_function = partial(self.split_by_regex, pattern=split_regex)
+
+        elif split_by == "page":
+            splitter_function = lambda text: self.split_by_regex(text=text, pattern="\f")
+
+        elif split_by == "paragraph":
+            splitter_function = lambda text: self.split_by_regex(text=text, pattern="\n\n")
+
+        elif split_by == "sentence":
+            splitter_function = lambda text: self.split_by_sentence_tokenizer(text=text)
+
+        elif split_by == "word":
+            splitter_function = lambda text: self.split_by_regex(text=text, pattern="\s+")
+
+        elif split_by == "token":
+            if isinstance(self.tokenizer, TokenizerI):
+                splitter_function = lambda text: self.split_by_sentence_tokenizer(text=text)
+            else:
+                splitter_function = lambda text: self.split_by_dense_tokenizer(text=text)
+        else:
+            raise ValueError("split_by must be either 'character', 'word', 'sentence', 'paragraph', 'page' or 'regex'")
+
+        final_documents = []
+        for document in documents:
+
+            # Split them into single unit documents
+            split_documents = self.split_into_units(
+                document=document, units=splitter_function(text=document.content), add_page_number=add_page_number
+            )[0]
+
+            # If we need to count the tokens, split it by token
+            if max_tokens:
+                for doc in split_documents:
+                    tokens = self.split_by_dense_tokenizer(text=doc.content)[0]
+                    doc.meta["tokens_count"] = len(tokens)
```

Comment:
Good point! I will consider some alternatives. Maybe I can make it private and call it `_tokens_count`, or `_preprocessing_tokens_count`? Or maybe is `DocumentMerger` that should know how to count tokens? Right now I use the Splitter for it, but it might be a good idea to make the Merger able to count tokens on its own. I'll try to avoid a circular dependency between Merger and Splitter on this feature.